scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers.js

```diff
@@ -0,0 +1,200 @@
+        body,
+        `${ruleNamespace}:${context.recipeName}/${context.ingredient}/expected-heading`
+      );
+      message.fatal = true;
+      logMissingIngredient(file, context);
+      return;
+    }
+
+    const nextHeading = select(`#${id} ~ h2`, tree);
+    const sectionStart = body.children.indexOf(heading);
+    const sectionEnd =
+      nextHeading !== null
+        ? body.children.indexOf(nextHeading)
+        : body.children.length;
+    const subTree = {
+      type: "root",
+      children: body.children.slice(sectionStart, sectionEnd)
+    };
+
+    let macroCount = 0;
+    visit(
+      subTree,
+      node => isMacro(node, "Compat"),
+      () => {
+        macroCount += 1;
+      }
+    );
+
+    if (macroCount !== 1) {
+      logMissingIngredient(file, context);
+    }
+  },
+  "data.examples": requireTopLevelHeading("Examples"),
+  "data.specifications": (tree, file, context) => {
+    const id = "Specifications";
+    const body = select(`body`, tree);
+
+    const heading = select(`h2#${id}`, tree);
+    if (heading === null) {
+      const message = file.message(
+        `Expected h2#${id} ${context.recipeName}: ${context.ingredient}`,
+        body,
+        `${ruleNamespace}:${context.recipeName}/${context.ingredient}/expected-heading`
+      );
+      message.fatal = true;
+      logMissingIngredient(file, context);
+      return;
+    }
+
+    const sectionStart = body.children.indexOf(heading);
+    const sectionEnd = body.children.indexOf(select(`#${id} ~ h2`, tree));
+    const subTree = {
+      type: "root",
+      children: body.children.slice(sectionStart, sectionEnd)
+    };
+
+    let sectionOk = false;
+    visit(subTree, "text", node => {
+      if (isMacro(node, "SpecName")) {
+        sectionOk = true;
+        return visit.SKIP;
+      }
+
+      if (node.value.includes("Not part of any standard")) {
+        sectionOk = true;
+        return visit.SKIP;
+      }
+    });
+
+    if (!sectionOk) {
+      const message = file.message(
+        `Expected SpecName macro for ${context.recipeName}: ${context.ingredient}`,
+        heading,
+        `${ruleNamespace}:${context.recipeName}/${context.ingredient}/expected-macro`
+      );
+      message.fatal = true;
+      logMissingIngredient(file, context);
+    }
+  },
+  "prose.description": requireTopLevelHeading("Description"),
+  "prose.error_type": requireTopLevelHeading("Error_type"),
+  "prose.message": requireTopLevelHeading("Message"),
+  "prose.see_also": requireTopLevelHeading("See_also"),
+  "prose.short_description": (tree, file, context) => {
+    // Get the page body without macro calls
+    const subTree = filter(select("body", tree), node => !isMacro(node));
+
+    // Drop any part of the subtree that's after the first H2
+    const sectionStart = 0;
+    const sectionEnd = subTree.children.indexOf(select("h2", subTree));
+    subTree.children = subTree.children.slice(sectionStart, sectionEnd);
```

Comment:
This is buggy because if there are no H2 elements, `sectionEnd` is -1 so `slice()` removes the last element.

Finding `short_description` is tricky and will probably take some tuning. For JS pages, it's probably something like: "there is a `<p>` element before the interactive example or an H2". We could also exclude `<p class="note"` and `<p class="warning"`.

Reply:
I've tried tuning the `short_description` handler some more and explaining my assumptions in a comment. There might be more ways in which it could be adjusted (is text content in a `<div>` OK? right now it's not), but it's pretty complicated as it is. In the interest of preventing this PR from growing larger, if you have any non-blocking feedback on the new `short_description` handler, please mention it, but I'll address them in a separate PR.